blockchain/stakeversion.go

```diff
@@ -0,0 +1,361 @@
+	node, err := b.findStakeVersionPriorNode(prevNode)
+	if err != nil {
+		return false
+	}
+	if node == nil {
+		return 0 >= minVer
+	}
+
+	// Tally both the total number of votes in the previous stake version validation
+	// interval and how many of those votes are at least the requested minimum
+	// version.
+	totalVotesFound := int32(0)
+	versionCount := int32(0)
+	iterNode := node
+	for i := int64(0); i < b.chainParams.StakeVersionInterval && iterNode != nil; i++ {
+		totalVotesFound += int32(len(iterNode.voterVersions))
+		for _, version := range iterNode.voterVersions {
+			if version >= minVer {
+				versionCount += 1
+			}
+		}
+
+		var err error
+		iterNode, err = b.getPrevNodeFromNode(iterNode)
+		if err != nil {
+			return false
+		}
+	}
+
+	// Determine the required amount of votes to reach supermajority.
+	numRequired := totalVotesFound * b.chainParams.StakeMajorityMultiplier /
+		b.chainParams.StakeMajorityDivisor
+
+	return versionCount >= numRequired
+}
+
+// isStakeMajorityVersion determines if minVer requirement is met based on
+// prevNode.  The function always uses the stake versions of the prior window.
+// For example, if StakeVersionInterval = 11 and StakeValidationHeight = 13 the
+// windows start at 13 + (11 * 2) 25 and are as follows: 24-34, 35-45, 46-56 ...
+// If height comes in at 35 we use the 24-34 window, up to height 45.
+// If height comes in at 46 we use the 35-45 window, up to height 56 etc.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) isStakeMajorityVersion(minVer uint32, prevNode *blockNode) bool {
+	// Walk blockchain backwards to calculate version.
+	node, err := b.findStakeVersionPriorNode(prevNode)
+	if err != nil {
+		return false
+	}
+	if node == nil {
+		return 0 >= minVer
+	}
+
+	// Calculate version.
+	versionCount := int32(0)
+	iterNode := node
+	for i := int64(0); i < b.chainParams.StakeVersionInterval && iterNode != nil; i++ {
+		if iterNode.header.StakeVersion >= minVer {
+			versionCount += 1
+		}
+
+		var err error
+		iterNode, err = b.getPrevNodeFromNode(iterNode)
+		if err != nil {
+			return false
+		}
+	}
+
+	// Determine the required amount of votes to reach supermajority.
+	numRequired := int32(b.chainParams.StakeVersionInterval) *
+		b.chainParams.StakeMajorityMultiplier /
+		b.chainParams.StakeMajorityDivisor
+
+	return versionCount >= numRequired
+}
+
+// calcPriorStakeVersion calculates the header stake version of the prior
+// interval.  The function walks the chain backwards by one interval and then
+// it performs a standard majority calculation.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) calcPriorStakeVersion(prevNode *blockNode) (uint32, error) {
+	// Walk blockchain backwards to calculate version.
+	node, err := b.findStakeVersionPriorNode(prevNode)
+	if err != nil {
+		return 0, err
+	}
+	if node == nil {
+		return 0, nil
+	}
+
+	// Calculate version.
```

Comment:
Same deal here.  This isn't calculating the version, rather it's tallying how many of each version the headers in the stake version interval have.

Perhaps:

```Go
// Tally how many of each stake version the block headers in the previous stake
// version validation interval have.
```